lib/Serialization/Serialization.cpp

```diff
@@ -4390,68 +4263,64 @@ void Serializer::writeAllDeclsAndTypes() {
   do {
     // Each of these loops can trigger the others to execute again, so repeat
     // until /all/ of the pending lists are empty.
-    while (!DeclsAndTypesToWrite.empty()) {
-      auto next = DeclsAndTypesToWrite.front();
-      DeclsAndTypesToWrite.pop();
+    while (DeclsToSerialize.hasMoreToSerialize()) {
+      auto next = DeclsToSerialize.popNext(Out.GetCurrentBitNo());
+      writeDecl(next);
```

Comment:
It would be nice to incorporate these repetitive loops into `Serialized` somehow. For instance, maybe you could rename `writeDecl`, `writeType`, etc. to just `writeEntity`, overloaded by parameter type, and then write:

```cpp
/// Return value: True if this call did no work; false otherwise.
template <typename RecordWriter>
bool Serialized::writeMore(RecordWriter &writer, llvm::BitstreamWriter &Out) {
  if (!hasMoreToSerialize())
    return true;

  do {
    auto next = popNext(Out.GetCurrentBitNo());
    writer.writeEntity(next);
  } while (hasMoreToSerialize());

  return false;
}
```

With that, the loop in `writeAllDeclsAndTypes()` is just:

```cpp
bool quiescent;
do {
  quiescent  = DeclsToSerialize.writeMore(*this, Out);
  quiescent &= TypesToSerialize.writeMore(*this, Out);
  quiescent &= LocalDeclContextsToSerialize.writeMore(*this, Out);
  // ...and so on...
} while (!quiescent);
```

Reply:
Ooh, clever. I'll look into it tomorrow.

Reply:
Neat idea :). Nitpick: if "writeMore" returns "true", I would expect that it means "yes, I wrote more stuff" but we're doing the opposite.

Reply:
I struggled a bit with the name for `quiescent`, which is why the method returns the value it does. In any case, Jordan can figure it out. 😀

Reply:
I'll come back to the full-on refactoring in a follow-up PR; it would mean eliminating the ambiguity between DeclContextsToSerialize and LocalDeclContextsToSerialize.